src/java.base/share/classes/sun/security/ssl/SSLSessionImpl.java

```diff
@@ -1456,6 +1562,259 @@ public List<SNIServerName> getRequestedServerNames() {
+        }
+
+        // Calculations are primarily based on protocol version.
+        switch (protocolVersion) {
+        case TLS13:  // HKDF-based
+
+            // Check the label/context lengths:
+            //       struct {
+            //           uint16 length = Length;
+            //           opaque label<7..255> = "tls13 " + Label;
+            //           opaque context<0..255> = Context;
+            //       } HkdfLabel;
+            // label can have 249 bytes (+6 for "tls13 "), and context 255
+
+            // RFC 8446 allows for length of 2^16-1 (65536), but RFC 5869
+            // states:
+            //
+            //     L    length of output keying material in octets
+            //          (<= 255*HashLen)
+            if (length >= (255 * cipherSuite.hashAlg.hashLength )) {
+                throw new IllegalArgumentException(
+                        "length is too large");
+            }
+
+            byte[] hkdfInfoLabel =
+                        ("tls13 " + label).getBytes(StandardCharsets.UTF_8);
+            if ((hkdfInfoLabel.length < 7) || hkdfInfoLabel.length > 255) {
+                throw new IllegalArgumentException(
+                        "label length outside range");
+            }
+
+            // If no context (null) is provided, RFC 8446 requires an empty
+            // context be used, unlike RFC 5705.
+            context = (context != null ? context : new byte[0]);
+            if (context.length > 255) {
+                throw new IllegalArgumentException(
+                        "context length outside range");
+            }
+
+            // Unlikely, but check anyway.
+            if (exporterMasterSecret == null) {
+                throw new RuntimeException(
+                        "Exporter master secret not captured");
+            }
+
+            // Do RFC 8446:7.1-7.5 calculations
+
+            /*
+             * TLS-Exporter(label, context_value, key_length) =
+             *     HKDF-Expand-Label(Derive-Secret(Secret, label, ""),
+             *         "exporter", Hash(context_value), key_length)
+             *
+             * Derive-Secret(Secret, Label, Messages) =
+             *     HKDF-Expand-Label(Secret, Label,
+             *         Transcript-Hash(Messages), Hash.length)
+             */
+
+            try {
+                // Use the ciphersuite's hashAlg for these calcs.
+                HashAlg hashAlg = cipherSuite.hashAlg;
+                KDF hkdf = KDF.getInstance(hashAlg.hkdfAlgorithm);
+
+                // First calculate the inner Derive-Secret(Secret, label, "")
+                MessageDigest md;
+                byte[] emptyHash;
+
+                // Create the "" digest...
+                try {
+                    md = MessageDigest.getInstance(hashAlg.toString());
+                    emptyHash = md.digest();
+                } catch (NoSuchAlgorithmException nsae) {
+                    throw new RuntimeException(
```

Comment:
We usually throw a `ProviderException` here, when some weird configuration removes a basic hash algorithm.

Reply:
I can do a `ProviderException` if you like, but our JSSE code is all over the place.  Sometimes it's a `RuntimeException`, other times we wrap it into into a `SSLException`, and sometimes we directly do a `fatal()`.  I'd say we do `RuntimeException` a little more often than `ProviderException`.

Reply:
Updated to to `ProviderException` since the point does fit the description of the class better.